### ISSUE TITLE
line/circle

### ISSUE THREAD

**Issue by Lin-Rolover** [jeffThompson/CollisionDetection#15]
Why do this?
float len = sqrt( (distX*distX) + (distY*distY) );
And after this
float dot = ( ((cx-x1)*(x2-x1)) + ((cy-y1)*(y2-y1)) ) / pow(len,2);

You have scalar aa
float len2 = (distX*distX) + (distY*distY);
And 
float dot = ( ((cx-x1)*(x2-x1)) + ((cy-y1)*(y2-y1)) ) / len2;

But what is that ? I am understand projection vector in other vector a*b/|b|. But what is : a*b/|b|*|b|?

**Comment by jeffThompson** [comment]
Thanks for noting this. It's been a looooong time since I researched this and my vector math isn't so good. I'm not sure I'm understanding you're question though?

**Comment by Lin-Rolover** [comment]
I found out. This is really a projection of a vector onto a vector, but also finding a collinear vector

1. 𝑝𝑟𝑥 (𝑌) = |𝑌| ∙ 𝑐𝑜𝑠(∠𝑋, 𝑌) = (𝑋∙ 𝑌) / |𝑋|  this number, the length of the projection with a sign (+, -)
2. 𝑃𝑅𝑋 (𝑌) = ((𝑋∙ 𝑌) / |𝑋|) ∙ (𝑋/|𝑋|) = ((𝑋,𝑌) ∙ 𝑋) / |𝑋|∙|𝑋|  - is a vector parallel (collinear) to the vector 𝑋.
and  |𝑋|∙|𝑋| <=>(𝑋∙𝑋)                => ((𝑋∙ 𝑌)/(𝑋∙𝑋)) ∙ 𝑋

ab/|b|*|b|   =>   ((𝑋∙ 𝑌)/(𝑋∙𝑋)) = dot   => dot  ∙  𝑋  - But this is a vector

Then you find the points
x = x1 + (dot ∙ (x2 - x1) )
y = y1 + (dot ∙ (y2 - y1) )

**Comment by jeffThompson** [comment]
Sorry, I think you lost me! Does this mean the code on the site isn't working correctly or could be simplified?

**Comment by Lin-Rolover** [comment]
Honestly, I haven’t tested your code at all except on the page at the very top. I just read and took for an idea, a solution algorithm, for example, what steps, what to watch for. I really liked the point / polygon algorithm, I took the comparison logic, and did not analyze it. 

In some solutions, I went from the opposite. For example, the intersection of a point and a line. I did not look for the intersection of a point with a line. I checked whether the point of the line belongs. Got this result

`pointLine (px,py, l) {`
`l = l.arrVert;`
`// if the scalar product of vectors <= 0, the point belongs to the segment`
`	return (`
`		(l[0] - px) * ( l[2] - px) + `
`		(l[1] - py) *  (l[3] - py) <= 0`
`	) ? {x:px, y:py} : false;`
`}`
All

If  think can put a buffer in, but now me no needed pointer
And my line / line algorithm is not very scary. Sorry.

But once again I want to thank you for great ideas
The code can be simplified, without doing unnecessary actions
